lib/cli-engine.js

```diff
@@ -372,31 +398,49 @@ CLIEngine.prototype = {
     executeOnFiles: function(files) {
 
         var results = [],
-            processed = [],
+            processed = {},
             options = this.options,
             configHelper = new Config(options),
-            ignoredPaths = IgnoredPaths.load(options),
-            exclude = ignoredPaths.contains.bind(ignoredPaths),
+            ignoredPaths = options.ignore ? IgnoredPaths.load(options).patterns : [],
+            extensions = options.extensions.map(function(ext) {
+                return ext.charAt(0) === "." ? ext : "." + ext;
+            }),
+            globOptions,
             stats;
 
-        traverse({
-            files: files,
-            extensions: options.extensions,
-            exclude: options.ignore ? exclude : false
-        }, function(filename) {
+        files = files.map(processPath);
+        ignoredPaths = ignoredPaths.map(processPath);
+
+        globOptions = {
+            nodir: true,
+            realpath: true,
+            ignore: ignoredPaths
+        };
+
+        function executeOnFile(filename) {
+            if (extensions.indexOf(path.extname(filename)) === -1 || processed[filename]) {
+                return;
+            }
 
             debug("Processing " + filename);
 
-            processed.push(filename);
+            processed[filename] = true;
             results.push(processFile(filename, configHelper));
+        }
+
+        files.forEach(function(pattern) {
+            glob.sync(pattern, globOptions).forEach(executeOnFile);
         });
 
         // only warn for files explicitly passed on the command line
         if (options.ignore) {
             files.forEach(function(file) {
-                if (fs.statSync(path.resolve(file)).isFile() && processed.indexOf(file) === -1) {
-                    results.push(createIgnoreResult(file));
-                }
+                try {
+                    var fullPath = path.resolve(file);
+                    if (fs.statSync(fullPath).isFile() && !processed[fullPath]) {
+                        results.push(createIgnoreResult(file));
+                    }
+                } catch (err) { /* ignore */ }
```

Comment:
Not a big fan of ignoring thrown errors. Under what situation will the error be thrown?